<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ ieugwasr</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ieugwasr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a></li>
    <li><a class="dropdown-item" href="../articles/local_ld.html">Running local LD operations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MRCIEU/ieugwasr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="afl2_chrpos.html">afl2_chrpos()</a></code>

        </dt>
        <dd>Look up allele frequencies and LD scores for 1000 genomes populations by chrpos</dd>
      </dl><dl><dt>

          <code><a href="afl2_list.html">afl2_list()</a></code>

        </dt>
        <dd>Retrieve a allele frequency and LD scores for pre-defined lists of variants</dd>
      </dl><dl><dt>

          <code><a href="afl2_rsid.html">afl2_rsid()</a></code>

        </dt>
        <dd>Look up allele frequencies and LD scores for 1000 genomes populations by rsid</dd>
      </dl><dl><dt>

          <code><a href="api_query.html">api_query()</a></code>

        </dt>
        <dd>Wrapper for sending queries and payloads to API</dd>
      </dl><dl><dt>

          <code><a href="api_status.html">api_status()</a></code>

        </dt>
        <dd>OpenGWAS server status</dd>
      </dl><dl><dt>

          <code><a href="associations.html">associations()</a></code>

        </dt>
        <dd>Query specific variants from specific GWAS</dd>
      </dl><dl><dt>

          <code><a href="batch_from_id.html">batch_from_id()</a></code>

        </dt>
        <dd>Extract batch name from study ID</dd>
      </dl><dl><dt>

          <code><a href="batches.html">batches()</a></code>

        </dt>
        <dd>Get list of data batches in IEU OpenGWAS database</dd>
      </dl><dl><dt>

          <code><a href="check_access_token.html">check_access_token()</a></code>

        </dt>
        <dd>Check if authentication has been made</dd>
      </dl><dl><dt>

          <code><a href="check_reset.html">check_reset()</a></code>

        </dt>
        <dd>Check if OpenGWAS allowance needs to be reset</dd>
      </dl><dl><dt>

          <code><a href="editcheck.html">editcheck()</a></code>

        </dt>
        <dd>Check datasets that are in process of being uploaded</dd>
      </dl><dl><dt>

          <code><a href="fill_n.html">fill_n()</a></code>

        </dt>
        <dd>Look up sample sizes when meta data is missing from associations</dd>
      </dl><dl><dt>

          <code><a href="get_opengwas_jwt.html">get_opengwas_jwt()</a></code>

        </dt>
        <dd>Retrieve OpenGWAS JSON Web Token from .Renviron file</dd>
      </dl><dl><dt>

          <code><a href="get_query_content.html">get_query_content()</a></code>

        </dt>
        <dd>Parse out json response from httr object</dd>
      </dl><dl><dt>

          <code><a href="gwasinfo.html">gwasinfo()</a></code>

        </dt>
        <dd>Get list of studies with available GWAS summary statistics through API</dd>
      </dl><dl><dt>

          <code><a href="gwasinfo_files.html">gwasinfo_files()</a></code>

        </dt>
        <dd>Get list of download URLs for each file associated with a dataset through API</dd>
      </dl><dl><dt>

          <code><a href="infer_ancestry.html">infer_ancestry()</a></code>

        </dt>
        <dd>Infer ancestry of GWAS dataset by matching against 1000 genomes allele frequencies</dd>
      </dl><dl><dt>

          <code><a href="ld_clump.html">ld_clump()</a></code>

        </dt>
        <dd>Perform LD clumping on SNP data</dd>
      </dl><dl><dt>

          <code><a href="ld_clump_api.html">ld_clump_api()</a></code>

        </dt>
        <dd>Perform clumping on the chosen variants using through API</dd>
      </dl><dl><dt>

          <code><a href="ld_clump_local.html">ld_clump_local()</a></code>

        </dt>
        <dd>Wrapper for clump function using local plink binary and ld reference dataset</dd>
      </dl><dl><dt>

          <code><a href="ld_matrix.html">ld_matrix()</a></code>

        </dt>
        <dd>Get LD matrix for list of SNPs</dd>
      </dl><dl><dt>

          <code><a href="ld_matrix_local.html">ld_matrix_local()</a></code>

        </dt>
        <dd>Get LD matrix using local plink binary and reference dataset</dd>
      </dl><dl><dt>

          <code><a href="ld_reflookup.html">ld_reflookup()</a></code>

        </dt>
        <dd>Check which rsids are present in a remote LD reference panel</dd>
      </dl><dl><dt>

          <code><a href="legacy_ids.html">legacy_ids()</a></code>

        </dt>
        <dd>Convert current IDs to legacy IDs</dd>
      </dl><dl><dt>

          <code><a href="logging_info.html">logging_info()</a></code>

        </dt>
        <dd>Details of how access token logs are used</dd>
      </dl><dl><dt>

          <code><a href="phewas.html">phewas()</a></code>

        </dt>
        <dd>Perform fast phewas of a specific variants against all available GWAS datasets</dd>
      </dl><dl><dt>

          <code><a href="print.ApiStatus.html">print(<i>&lt;ApiStatus&gt;</i>)</a></code>

        </dt>
        <dd>Print API status</dd>
      </dl><dl><dt>

          <code><a href="print.GwasInfo.html">print(<i>&lt;GwasInfo&gt;</i>)</a></code>

        </dt>
        <dd>Print GWAS information</dd>
      </dl><dl><dt>

          <code><a href="select_api.html">select_api()</a></code>

        </dt>
        <dd>Toggle API address between development and release</dd>
      </dl><dl><dt>

          <code><a href="set_reset.html">set_reset()</a></code>

        </dt>
        <dd>Set the reset time for OpenGWAS allowance</dd>
      </dl><dl><dt>

          <code><a href="tophits.html">tophits()</a></code>

        </dt>
        <dd>Obtain top hits from a GWAS dataset</dd>
      </dl><dl><dt>

          <code><a href="user.html">user()</a></code>

        </dt>
        <dd>Get user details</dd>
      </dl><dl><dt>

          <code><a href="variants_chrpos.html">variants_chrpos()</a></code>

        </dt>
        <dd>Obtain information about chr pos and surrounding region</dd>
      </dl><dl><dt>

          <code><a href="variants_gene.html">variants_gene()</a></code>

        </dt>
        <dd>Obtain variants around a gene</dd>
      </dl><dl><dt>

          <code><a href="variants_rsid.html">variants_rsid()</a></code>

        </dt>
        <dd>Obtain information about rsid</dd>
      </dl><dl><dt>

          <code><a href="variants_to_rsid.html">variants_to_rsid()</a></code>

        </dt>
        <dd>Convert mixed array of rsid and chrpos to list of rsid</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Ben Elsworth, Tom Palmer, Rita Rasteiro.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

