<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Perform fast queries against a massive database of complete GWAS summary data • ieugwasr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Perform fast queries against a massive database of complete GWAS summary data">
<meta property="og:description" content="ieugwasr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ieugwasr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a>
    </li>
    <li>
      <a href="../articles/local_ld.html">Running local LD operations</a>
    </li>
    <li>
      <a href="../articles/timings.html">Comparison of rsid vs chr:pos lookup in ElasticSearch</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrcieu/ieugwasr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Perform fast queries against a massive database of complete GWAS summary data</h1>
                        <h4 class="author">Gibran Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mrcieu/ieugwasr/blob/master/vignettes/guide.Rmd"><code>vignettes/guide.Rmd</code></a></small>
      <div class="hidden name"><code>guide.Rmd</code></div>

    </div>

    
    
<p>The <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> comprises over 10,000 curated, QC’d and harmonised complete GWAS summary datasets and can be queried using an API. See <a href="http://gwasapi.mrcieu.ac.uk/docs/">here</a> for documentation on the API itself. This R package is a wrapper to make generic calls to the API, plus convenience functions for specific queries.</p>
<div id="authentication" class="section level2">
<h2 class="hasAnchor">
<a href="#authentication" class="anchor"></a>Authentication</h2>
<p>Most datasets in the database are public and don’t need authentication. But if you want to access a private dataset that is linked to your (gmail) email address, you need to authenticate the query using a method known as Google OAuth2.0.</p>
<p>Essentially - you run this command at the start of your session:</p>
<pre><code><a href="../reference/get_access_token.html">ieugwasr::get_access_token()</a></code></pre>
<p>which will open up a web browser asking you to provide your google username and password, and upon doing so a directory will be created in your working directory called <code>ieugwasr_oauth</code>. This directory contains a file that looks like this: <code>&lt;random_string&gt;_&lt;email@address&gt;</code>. It is a binary file (not human readable), which contains your access token, and it acts as a convenient way to hold a randomly generated password.</p>
<p>If you are using a server which doesn’t have a graphic user interface then the <code><a href="../reference/get_access_token.html">ieugwasr::get_access_token()</a></code> method is not going to work. You need to generate the <code>ieugwasr_oauth</code> directory and token file on a computer that has a web browser, and then copy that directory (containing the token file) to your server (to the relevant work directory).</p>
<p>If you are using R in a working directory that does not have write permissions then this command will fail, please navigate to a directory that does have write permissions.</p>
<p>If you need to run this in a non-interactive script then you can generate the token file on an interactive computer, copy that file to the working directory that R will be running from, and then run a batch (non-interactive).</p>
<p>You can test to see if you have authenticated using the function</p>
<pre><code><a href="../reference/check_access_token.html">ieugwasr::check_access_token()</a></code></pre>
<p>It will return <code>NULL</code> if you are not authenticated, or a long random token string if you are.</p>
<p>To unauthenticate, simply delete the relevant file in the <code>ieugwasr_oauth</code> folder, or delete the folder entirely.</p>
</div>
<div id="general-api-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#general-api-queries" class="anchor"></a>General API queries</h2>
<p>The API has a number of endpoints documented <a href="http://ieu-db-interface.epi.bris.ac.uk:8082/docs/">here</a>. A general way to access them in R is using the <code>api_query</code> function. There are two types of endpoints - <code>GET</code> and <code>POST</code>.</p>
<ul>
<li>
<code>GET</code> - you provide a single URL which includes the endpoint and query. For example, for the <code>association</code> endpoint you can obtain some rsids in some studies, e.g.
<ul>
<li><code><a href="../reference/api_query.html">api_query("associations/ieu-a-2,ieu-a-7/rs234,rs123")</a></code></li>
</ul>
</li>
<li>
<code>POST</code> - Here you send a “payload” to the endpoint. So, the path specifies the endpoint and you add a list of query specifications. This is useful for long lists of rsids being queried, for example
<ul>
<li><code><a href="../reference/api_query.html">api_query("associations", query=list(rsid=c("rs234", "rs123"), id=c("ieu-a-2", "ieu-a-7")))</a></code></li>
</ul>
</li>
</ul>
<p>The <code>api_query</code> function returns a <code>response</code> object from the <code>httr</code> package. See below for a list of functions that make the input and output to <code>api_query</code> more convenient.</p>
</div>
<div id="get-api-status" class="section level2">
<h2 class="hasAnchor">
<a href="#get-api-status" class="anchor"></a>Get API status</h2>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="../reference/api_status.html">api_status</a></span>()</pre></body></html></div>
</div>
<div id="get-list-of-all-available-studies" class="section level2">
<h2 class="hasAnchor">
<a href="#get-list-of-all-available-studies" class="anchor"></a>Get list of all available studies</h2>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="../reference/gwasinfo.html">gwasinfo</a></span>()</pre></body></html></div>
</div>
<div id="get-list-of-a-specific-study" class="section level2">
<h2 class="hasAnchor">
<a href="#get-list-of-a-specific-study" class="anchor"></a>Get list of a specific study</h2>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="../reference/gwasinfo.html">gwasinfo</a></span>(<span class="st">"ieu-a-2"</span>)</pre></body></html></div>
</div>
<div id="extract-particular-associations-from-particular-studies" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-particular-associations-from-particular-studies" class="anchor"></a>Extract particular associations from particular studies</h2>
<p>Provide a list of variants to be obtained from a list of studies:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="../reference/associations.html">associations</a></span>(<span class="kw">variants</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rs123"</span>, <span class="st">"7:105561135"</span>), <span class="kw">id</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ieu-a-2"</span>, <span class="st">"ieu-a-7"</span>))</pre></body></html></div>
<p>By default this will look for LD proxies using 1000 genomes reference data (Europeans only, the reference panel has INDELs removed and only retains SNPs with MAF &gt; 0.01). This behaviour can be turned off using <code>proxies=0</code> as an argument.</p>
<p>Note that the queries are performed on rsids, but chromosome:position values will be automatically converted. A range query can be done using e.g.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="../reference/associations.html">associations</a></span>(<span class="kw">variants</span><span class="kw">=</span><span class="st">"7:105561135-105563135"</span>, <span class="kw">id</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ieu-a-2"</span>), <span class="kw">proxies</span><span class="kw">=</span><span class="fl">0</span>)</pre></body></html></div>
</div>
<div id="get-the-tophits-from-a-study" class="section level2">
<h2 class="hasAnchor">
<a href="#get-the-tophits-from-a-study" class="anchor"></a>Get the tophits from a study</h2>
<p>The tophits can be obtained using</p>
<pre><code><a href="../reference/tophits.html">tophits(id="ieu-a-2")</a></code></pre>
<p>Note that it will perform strict clumping by default (r2 = 0.001 and radius = 10000kb). This can be turned off with <code>clump=0</code>.</p>
</div>
<div id="perform-phewas" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-phewas" class="anchor"></a>Perform PheWAS</h2>
<p>Lookup association of specified variants across every study, returning at a particular threshold. Note that no LD proxy lookups are made here.</p>
<pre><code><a href="../reference/phewas.html">phewas(variants="rs1205", pval=1e-5)</a></code></pre>
<p>PheWAS can also be performed in only specific subsets of the data. The datasets in the IGD are organised by batch, you can see info about it here: <a href="https://gwas.mrcieu.ac.uk/datasets/" class="uri">https://gwas.mrcieu.ac.uk/datasets/</a> or get a list of batches and their descriptions using:</p>
<pre><code><a href="../reference/batches.html">batches()</a></code></pre>
<p>You can perform PheWAS in only specified batches using:</p>
<pre><code><a href="../reference/phewas.html">phewas(variants="rs1205", pval=1e-5, batch=c('ieu-a', 'ukb-b'))</a></code></pre>
<p>By default PheWAS is performed in all batches (which is of course somewhat slower).</p>
</div>
<div id="ld-clumping" class="section level2">
<h2 class="hasAnchor">
<a href="#ld-clumping" class="anchor"></a>LD clumping</h2>
<p>The API has a wrapper around <a href="https://www.cog-genomics.org/plink/1.9">plink version 1.90</a> and can use it to perform clumping with an LD reference panel from 1000 genomes reference data.</p>
<pre><code>a &lt;- tophits(id="ieu-a-2", clump=0)
b &lt;- ld_clump(
    dplyr::tibble(rsid=a$name, pval=a$p, id=a$id)
)</code></pre>
<p>There are 5 super-populations that can be requested via the <code>pop</code> argument. By default this will use the Europeans subset (EUR super-population). The reference panel has INDELs removed and only retains SNPs with MAF &gt; 0.01 in the selected population.</p>
<p>Note that you can perform the same operation locally if you provide a path to plink and a bed/bim/fam LD reference dataset. e.g.</p>
<pre><code><a href="../reference/ld_clump.html">ld_clump(
    dplyr::tibble(rsid=a$name, pval=a$p, id=a$id),
    plink_bin = "/path/to/plink",
    bfile = "/path/to/reference_data"
)</a></code></pre>
<p>See the following vignette for more information: <a href="local_ld.html">Running local LD operations</a></p>
</div>
<div id="ld-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#ld-matrix" class="anchor"></a>LD matrix</h2>
<p>Similarly, a matrix of LD r values can be generated using</p>
<pre><code><a href="../reference/ld_matrix.html">ld_matrix(b$variant)</a></code></pre>
<p>This uses the API by default but is limited to only 500 variants. You can use, instead, local plink and LD reference data in the same manner as in the <code>ld_clump</code> function, e.g.</p>
<pre><code><a href="../reference/ld_matrix.html">ld_matrix(b$variant, plink_bin = "/path/to/plink", bfile = "/path/to/reference_data")</a></code></pre>
<p>There are 5 super-populations that can be requested via the <code>pop</code> argument. By default this will use the Europeans subset (EUR super-population). The reference panel has INDELs removed and only retains SNPs with MAF &gt; 0.01 in the selected population.</p>
<p>See the following vignette for more information: <a href="local_ld.html">Running local LD operations</a></p>
</div>
<div id="variant-information" class="section level2">
<h2 class="hasAnchor">
<a href="#variant-information" class="anchor"></a>Variant information</h2>
<p>Translating between rsids and chromosome:position, while also getting other information, can be achieved.</p>
<p>The <code>chrpos</code> argument can accept the following</p>
<ul>
<li><code>&lt;chr&gt;:&lt;position&gt;</code></li>
<li><code>&lt;chr&gt;:&lt;start&gt;-&lt;end&gt;</code></li>
</ul>
<p>For example</p>
<pre><code>a &lt;- variants_chrpos(c("7:105561135-105563135", "10:44865737"))</code></pre>
<p>This provides a table with dbSNP variant IDs, gene info, and various other metadata. Similar data can be obtained from searching by rsid</p>
<pre><code>b &lt;- variants_rsid(c("rs234", "rs333"))</code></pre>
<p>And a list of variants within a particular gene region can also be found. Provide a ensembl or entrez gene ID (e.g. ENSG00000123374 or 1017) to the following:</p>
<pre><code>c &lt;- variants_gene("ENSG00000123374")</code></pre>
</div>
<div id="extracting-gwas-summary-data-based-on-gene-region" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-gwas-summary-data-based-on-gene-region" class="anchor"></a>Extracting GWAS summary data based on gene region</h2>
<p>Here is an example of how to obtain summary data for some datasets for a gene region. As an example, we’ll extract CDK2 (HGNC number 1017) from a BMI dataset (ieu-a-2)</p>
<p>Use the <a href="https://bioconductor.org/packages/release/bioc/html/mygene.html">mygene</a> bioconductor package to query the <a href="https://mygene.info/">mygene.info</a> API.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mygene</span>)
<span class="no">a</span> <span class="kw">&lt;-</span> <span class="kw pkg">mygene</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/mygene/man/getGene.html">getGene</a></span>(<span class="st">"1017"</span>, <span class="kw">fields</span><span class="kw">=</span><span class="st">"genomic_pos_hg19"</span>)
<span class="no">r</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">a</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">genomic_pos_hg19</span>$<span class="no">chr</span>, <span class="st">":"</span>, <span class="no">a</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">genomic_pos_hg19</span>$<span class="no">start</span>, <span class="st">"-"</span>, <span class="no">a</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">genomic_pos_hg19</span>$<span class="no">end</span>)
<span class="no">b</span> <span class="kw">&lt;-</span> <span class="kw pkg">ieugwasr</span><span class="kw ns">::</span><span class="fu"><a href="../reference/associations.html">associations</a></span>(<span class="no">r</span>, <span class="st">"ieu-a-2"</span>)</pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
