<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ieugwasr">
<title>Perform fast queries against a massive database of complete GWAS summary data • ieugwasr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform fast queries against a massive database of complete GWAS summary data">
<meta property="og:description" content="ieugwasr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ieugwasr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2-9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a>
    <a class="dropdown-item" href="../articles/local_ld.html">Running local LD operations</a>
    <a class="dropdown-item" href="../articles/timings.html">Comparison of rsid vs chr:pos lookup in ElasticSearch</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MRCIEU/ieugwasr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform fast queries against a massive database of complete GWAS summary data</h1>
                        <h4 data-toc-skip class="author">Gibran
Hemani</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/ieugwasr/blob/HEAD/vignettes/guide.Rmd" class="external-link"><code>vignettes/guide.Rmd</code></a></small>
      <div class="d-none name"><code>guide.Rmd</code></div>
    </div>

    
    
<p>The <a href="https://gwas.mrcieu.ac.uk/" class="external-link">OpenGWAS database</a>
comprises over 50,000 curated, QC’d and harmonised complete GWAS summary
datasets and can be queried using an API. See <a href="https://api.opengwas.io/api/" class="external-link">here</a> for documentation on the
API itself. This R package is a wrapper to make generic calls to the
API, plus convenience functions for specific queries.</p>
<div class="section level2">
<h2 id="authentication">Authentication<a class="anchor" aria-label="anchor" href="#authentication"></a>
</h2>
<p>Most datasets in the database are public and don’t need
authentication. But if you want to access a private dataset that is
linked to your (gmail) email address, you need to authenticate the query
using a method known as Google OAuth2.0.</p>
<p>Essentially - you run this command at the start of your session:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/get_access_token.html">get_access_token</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>which will open up a web browser asking you to provide your google
username and password, and upon doing so a directory will be created in
your working directory called <code>ieugwasr_oauth</code>. This
directory contains a file that looks like this:
<code>&lt;random_string&gt;_&lt;email@address&gt;</code>. It is a binary
file (not human readable), which contains your access token, and it acts
as a convenient way to hold a randomly generated password.</p>
<p>If you are using a server which doesn’t have a graphic user interface
then the <code><a href="../reference/get_access_token.html">ieugwasr::get_access_token()</a></code> method is not going
to work. You need to generate the <code>ieugwasr_oauth</code> directory
and token file on a computer that has a web browser, and then copy that
directory (containing the token file) to your server (to the relevant
work directory).</p>
<p>If you are using R in a working directory that does not have write
permissions then this command will fail, please navigate to a directory
that does have write permissions.</p>
<p>If you need to run this in a non-interactive script then you can
generate the token file on an interactive computer, copy that file to
the working directory that R will be running from, and then run a batch
(non-interactive).</p>
<p>You can test to see if you have authenticated using the function</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/check_access_token.html">check_access_token</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>It will return <code>NULL</code> if you are not authenticated, or a
long random token string if you are.</p>
<p>To unauthenticate, simply delete the relevant file in the
<code>ieugwasr_oauth</code> folder, or delete the folder entirely.</p>
</div>
<div class="section level2">
<h2 id="general-api-queries">General API queries<a class="anchor" aria-label="anchor" href="#general-api-queries"></a>
</h2>
<p>The API has a number of endpoints documented <a href="https://api.opengwas.io/api/" class="external-link">here</a>. A general way to access
them in R is using the <code>api_query</code> function. There are two
types of endpoints - <code>GET</code> and <code>POST</code>.</p>
<ul>
<li>
<code>GET</code> - you provide a single URL which includes the
endpoint and query. For example, for the <code>association</code>
endpoint you can obtain some rsids in some studies, e.g.
<ul>
<li><code>api_query("associations/ieu-a-2,ieu-a-7/rs234,rs123")</code></li>
</ul>
</li>
<li>
<code>POST</code> - Here you send a “payload” to the endpoint. So,
the path specifies the endpoint and you add a list of query
specifications. This is useful for long lists of rsids being queried,
for example
<ul>
<li><code>api_query("associations", query=list(rsid=c("rs234", "rs123"), id=c("ieu-a-2", "ieu-a-7")))</code></li>
</ul>
</li>
</ul>
<p>The <code>api_query</code> function returns a <code>response</code>
object from the <code>httr</code> package. See below for a list of
functions that make the input and output to <code>api_query</code> more
convenient.</p>
</div>
<div class="section level2">
<h2 id="get-api-status">Get API status<a class="anchor" aria-label="anchor" href="#get-api-status"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/ieugwasr" class="external-link">ieugwasr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/api_status.html">api_status</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-list-of-all-available-studies">Get list of all available studies<a class="anchor" aria-label="anchor" href="#get-list-of-all-available-studies"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gwasinfo.html">gwasinfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-list-of-a-specific-study">Get list of a specific study<a class="anchor" aria-label="anchor" href="#get-list-of-a-specific-study"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gwasinfo.html">gwasinfo</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extract-particular-associations-from-particular-studies">Extract particular associations from particular studies<a class="anchor" aria-label="anchor" href="#extract-particular-associations-from-particular-studies"></a>
</h2>
<p>Provide a list of variants to be obtained from a list of studies:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span>variants<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs123"</span>, <span class="st">"7:105561135"</span><span class="op">)</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span>, <span class="st">"ieu-a-7"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>By default this will look for LD proxies using 1000 genomes reference
data (Europeans only, the reference panel has INDELs removed and only
retains SNPs with MAF &gt; 0.01). This behaviour can be turned off using
<code>proxies=0</code> as an argument.</p>
<p>Note that the queries are performed on rsids, but chromosome:position
values will be automatically converted. A range query can be done using
e.g.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span>variants<span class="op">=</span><span class="st">"7:105561135-105563135"</span>, id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span><span class="op">)</span>, proxies<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-the-tophits-from-a-study">Get the tophits from a study<a class="anchor" aria-label="anchor" href="#get-the-tophits-from-a-study"></a>
</h2>
<p>The tophits can be obtained using</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tophits.html">tophits</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"ieu-a-2"</span><span class="op">)</span></span></code></pre></div>
<p>Note that it will perform strict clumping by default (r2 = 0.001 and
radius = 10000kb). This can be turned off with <code>clump=0</code>.</p>
</div>
<div class="section level2">
<h2 id="perform-phewas">Perform PheWAS<a class="anchor" aria-label="anchor" href="#perform-phewas"></a>
</h2>
<p>Lookup association of specified variants across every study,
returning at a particular threshold. Note that no LD proxy lookups are
made here.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/phewas.html">phewas</a></span><span class="op">(</span>variants<span class="op">=</span><span class="st">"rs1205"</span>, pval<span class="op">=</span><span class="fl">1e-5</span><span class="op">)</span></span></code></pre></div>
<p>PheWAS can also be performed in only specific subsets of the data.
The datasets in the IGD are organised by batch, you can see info about
it here: <a href="https://gwas.mrcieu.ac.uk/datasets/" class="external-link uri">https://gwas.mrcieu.ac.uk/datasets/</a> or get a list of
batches and their descriptions using:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/batches.html">batches</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can perform PheWAS in only specified batches using:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/phewas.html">phewas</a></span><span class="op">(</span>variants<span class="op">=</span><span class="st">"rs1205"</span>, pval<span class="op">=</span><span class="fl">1e-5</span>, batch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ieu-a'</span>, <span class="st">'ukb-b'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>By default PheWAS is performed in all batches (which is of course
somewhat slower).</p>
</div>
<div class="section level2">
<h2 id="ld-clumping">LD clumping<a class="anchor" aria-label="anchor" href="#ld-clumping"></a>
</h2>
<p>The API has a wrapper around <a href="https://www.cog-genomics.org/plink/1.9" class="external-link">plink version 1.90</a> and
can use it to perform clumping with an LD reference panel from 1000
genomes reference data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tophits.html">tophits</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"ieu-a-2"</span>, clump<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ld_clump.html">ld_clump</a></span><span class="op">(</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>rsid<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">name</span>, pval<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">p</span>, id<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>There are 5 super-populations that can be requested via the
<code>pop</code> argument. By default this will use the Europeans subset
(EUR super-population). The reference panel has INDELs removed and only
retains SNPs with MAF &gt; 0.01 in the selected population.</p>
<p>Note that you can perform the same operation locally if you provide a
path to plink and a bed/bim/fam LD reference dataset. e.g.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_clump.html">ld_clump</a></span><span class="op">(</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>rsid<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">name</span>, pval<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">p</span>, id<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">id</span><span class="op">)</span>,</span>
<span>    plink_bin <span class="op">=</span> <span class="st">"/path/to/plink"</span>,</span>
<span>    bfile <span class="op">=</span> <span class="st">"/path/to/reference_data"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>See the following vignette for more information: <a href="local_ld.html">Running local LD operations</a></p>
</div>
<div class="section level2">
<h2 id="ld-matrix">LD matrix<a class="anchor" aria-label="anchor" href="#ld-matrix"></a>
</h2>
<p>Similarly, a matrix of LD r values can be generated using</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span><span class="va">b</span><span class="op">$</span><span class="va">variant</span><span class="op">)</span></span></code></pre></div>
<p>This uses the API by default but is limited to only 500 variants. You
can use, instead, local plink and LD reference data in the same manner
as in the <code>ld_clump</code> function, e.g.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span><span class="va">b</span><span class="op">$</span><span class="va">variant</span>, plink_bin <span class="op">=</span> <span class="st">"/path/to/plink"</span>, bfile <span class="op">=</span> <span class="st">"/path/to/reference_data"</span><span class="op">)</span></span></code></pre></div>
<p>There are 5 super-populations that can be requested via the
<code>pop</code> argument. By default this will use the Europeans subset
(EUR super-population). The reference panel has INDELs removed and only
retains SNPs with MAF &gt; 0.01 in the selected population.</p>
<p>Super-populations:</p>
<ul>
<li>EUR = European</li>
<li>EAS = East Asian</li>
<li>AMR = Admixed American</li>
<li>SAS = South Asian</li>
<li>AFR = African</li>
</ul>
<p>See the following vignette for more information: <a href="local_ld.html">Running local LD operations</a></p>
</div>
<div class="section level2">
<h2 id="variant-information">Variant information<a class="anchor" aria-label="anchor" href="#variant-information"></a>
</h2>
<p>Translating between rsids and chromosome:position, while also getting
other information, can be achieved.</p>
<p>The <code>chrpos</code> argument can accept the following</p>
<ul>
<li><code>&lt;chr&gt;:&lt;position&gt;</code></li>
<li><code>&lt;chr&gt;:&lt;start&gt;-&lt;end&gt;</code></li>
</ul>
<p>For example</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/variants_chrpos.html">variants_chrpos</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"7:105561135-105563135"</span>, <span class="st">"10:44865737"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This provides a table with dbSNP variant IDs, gene info, and various
other metadata. Similar data can be obtained from searching by rsid</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/variants_rsid.html">variants_rsid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs234"</span>, <span class="st">"rs333"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>And a list of variants within a particular gene region can also be
found. Provide a ensembl or entrez gene ID (e.g. ENSG00000123374 or
1017) to the following:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/variants_gene.html">variants_gene</a></span><span class="op">(</span><span class="st">"ENSG00000123374"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-gwas-summary-data-based-on-gene-region">Extracting GWAS summary data based on gene region<a class="anchor" aria-label="anchor" href="#extracting-gwas-summary-data-based-on-gene-region"></a>
</h2>
<p>Here is an example of how to obtain summary data for some datasets
for a gene region. As an example, we’ll extract CDK2 (HGNC number 1017)
from a BMI dataset (ieu-a-2)</p>
<p>Use the <a href="https://bioconductor.org/packages/release/bioc/html/mygene.html" class="external-link">mygene</a>
bioconductor package to query the <a href="https://mygene.info/" class="external-link">mygene.info</a> API.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mygene</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu">mygene</span><span class="fu">::</span><span class="fu">getGene</span><span class="op">(</span><span class="st">"1017"</span>, fields<span class="op">=</span><span class="st">"genomic_pos_hg19"</span><span class="op">)</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">a</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genomic_pos_hg19</span><span class="op">$</span><span class="va">chr</span>, <span class="st">":"</span>, <span class="va">a</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genomic_pos_hg19</span><span class="op">$</span><span class="va">start</span>, <span class="st">"-"</span>, <span class="va">a</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genomic_pos_hg19</span><span class="op">$</span><span class="va">end</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span><span class="va">r</span>, <span class="st">"ieu-a-2"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="genomes-annotations">1000 genomes annotations<a class="anchor" aria-label="anchor" href="#genomes-annotations"></a>
</h2>
<p>The OpenGWAS database contains a database of population annotations
from the 1000 genomes project - the alternative allele frequencies and
the LD scores for each variant, calculated for each super population
separately. Only variants are present if they are MAF &gt; 1% in at
least one super population. You can access this info in different
ways</p>
<ol style="list-style-type: decimal">
<li>
<p>Look up a particular set of rsids</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/afl2_rsid.html">afl2_rsid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs234"</span>, <span class="st">"rs123"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Look up a set of positions or regions</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/afl2_chrpos.html">afl2_chrpos</a></span><span class="op">(</span><span class="st">"1:100000-900000"</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Extract annotations for a list of 20k variants that are common in
all super populations, and evenly spaced across the genome</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/afl2_list.html">afl2_list</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Extract annotations for a 1.3 million HapMap3 variants</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/afl2_list.html">afl2_list</a></span><span class="op">(</span><span class="st">"hapmap3"</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Infer the ancestry of a particular study by comparing the allele
frequencies with different super population reference frequencies</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snplist</span> <span class="op">&lt;-</span> <span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/afl2_list.html">afl2_list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">eur_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span><span class="va">snplist</span><span class="op">$</span><span class="va">rsid</span>, <span class="st">"ieu-a-2"</span><span class="op">)</span></span>
<span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/infer_ancestry.html">infer_ancestry</a></span><span class="op">(</span><span class="va">eur_example</span>, <span class="va">snplist</span><span class="op">)</span></span>
<span><span class="va">eas_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span><span class="va">snplist</span><span class="op">$</span><span class="va">rsid</span>, <span class="st">"bbj-a-10"</span><span class="op">)</span></span>
<span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="../reference/infer_ancestry.html">infer_ancestry</a></span><span class="op">(</span><span class="va">eur_example</span>, <span class="va">snplist</span><span class="op">)</span></span></code></pre></div>
</li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gibran Hemani, Ben Elsworth, Tom Palmer, Rita Rasteiro.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
