<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ieugwasr">
<title>Running local LD operations • ieugwasr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running local LD operations">
<meta property="og:description" content="ieugwasr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ieugwasr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <a class="dropdown-item" href="../articles/guide.html">Perform fast queries against a massive database of complete GWAS summary data</a>
    <a class="dropdown-item" href="../articles/local_ld.html">Running local LD operations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MRCIEU/ieugwasr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running local LD operations</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MRCIEU/ieugwasr/blob/HEAD/vignettes/local_ld.Rmd" class="external-link"><code>vignettes/local_ld.Rmd</code></a></small>
      <div class="d-none name"><code>local_ld.Rmd</code></div>
    </div>

    
    
<p>We have tried to provide useful cloud-based functionality for many
operations, including relatively demanding LD operations. If you are
running a large number of LD operations, we request that you think about
performing those locally rather than through the API. We have tried to
write the software to enable this to work seamlessly. Some examples
below.</p>
<p>LD operations available on the OpenGWAS API</p>
<ul>
<li>LD clumping</li>
<li>Generating LD matrices</li>
<li>Looking for LD proxies</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MRCIEU/ieugwasr" class="external-link">ieugwasr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; OpenGWAS updates:</span></span>
<span><span class="co">#&gt;   Date: 2024-03-07</span></span>
<span><span class="co">#&gt;   [&gt;] There is exceptional load on the OpenGWAS servers.</span></span>
<span><span class="co">#&gt;   [&gt;] Urgent infrastructure development being performed.</span></span>
<span><span class="co">#&gt;   [&gt;] See local options for analysis: https://mrcieu.github.io/gwasvcf/.</span></span></code></pre></div>
<div class="section level2">
<h2 id="ld-clumping">LD clumping<a class="anchor" aria-label="anchor" href="#ld-clumping"></a>
</h2>
<p>The API has a wrapper around <a href="https://www.cog-genomics.org/plink/1.9" class="external-link">plink version 1.90</a> and
can use it to perform clumping with an LD reference panel from 1000
genomes reference data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tophits.html">tophits</a></span><span class="op">(</span>id<span class="op">=</span><span class="st">"ieu-a-2"</span>, clump<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ld_clump.html">ld_clump</a></span><span class="op">(</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>rsid<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">name</span>, pval<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">p</span>, id<span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>There are 5 super-populations that can be requested via the
<code>pop</code> argument. By default this will use the Europeans subset
(EUR super-population). The reference panel has INDELs removed and only
retains SNPs with MAF &gt; 0.01 in the selected population.</p>
<p>Note that you can perform the same operation locally if you provide a
path to plink and a bed/bim/fam LD reference dataset.</p>
<p>To get a path to plink you can do the following:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"explodecomputer/genetics.binaRies"</span><span class="op">)</span></span>
<span><span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu">get_plink_binary</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>To get the same LD reference dataset that is used by the API, you can
download it directly from here:</p>
<p><a href="http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz" class="external-link uri">http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz</a></p>
<p>This contains an LD reference panel for each of the 5
super-populations in the 1000 genomes reference dataset. e.g. for the
European super population it has the following files:</p>
<ul>
<li><code>EUR.bed</code></li>
<li><code>EUR.bim</code></li>
<li><code>EUR.fam</code></li>
</ul>
<p>Now supposing in R you have a dataframe, <code>dat</code>, with the
following columns:</p>
<ul>
<li><code>rsid</code></li>
<li><code>pval</code></li>
<li><code>trait_id</code></li>
</ul>
<p>to perform clumping, just do the following:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_clump.html">ld_clump</a></span><span class="op">(</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>rsid<span class="op">=</span><span class="va">dat</span><span class="op">$</span><span class="va">rsid</span>, pval<span class="op">=</span><span class="va">dat</span><span class="op">$</span><span class="va">pval</span>, id<span class="op">=</span><span class="va">dat</span><span class="op">$</span><span class="va">trait_id</span><span class="op">)</span>,</span>
<span>    plink_bin <span class="op">=</span> <span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu">get_plink_binary</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    bfile <span class="op">=</span> <span class="st">"/path/to/reference/EUR"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ld-matrix">LD matrix<a class="anchor" aria-label="anchor" href="#ld-matrix"></a>
</h2>
<p>Similarly, a matrix of LD r values can be generated using</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span><span class="va">b</span><span class="op">$</span><span class="va">variant</span><span class="op">)</span></span></code></pre></div>
<p>This uses the API by default but is limited to only 500 variants. You
can use, instead, local plink and LD reference data in the same manner
as in the <code>ld_clump</code> function, e.g.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ld_matrix.html">ld_matrix</a></span><span class="op">(</span></span>
<span>    <span class="va">dat</span><span class="op">$</span><span class="va">rsid</span>,</span>
<span>    plink_bin <span class="op">=</span> <span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu">get_plink_binary</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    bfile <span class="op">=</span> <span class="st">"/path/to/reference/EUR"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ld-proxies">LD proxies<a class="anchor" aria-label="anchor" href="#ld-proxies"></a>
</h2>
<p>To automatically extract variants from a dataset, and search for LD
proxies when a requested variant is not present in the dataset, please
look at the options available in the <a href="https://mrcieu.github.io/gwasvcf/" class="external-link">gwasvcf</a> package:</p>
<p><a href="https://mrcieu.github.io/gwasvcf/articles/guide.html#ld-proxies-1" class="external-link uri">https://mrcieu.github.io/gwasvcf/articles/guide.html#ld-proxies-1</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Ben Elsworth, Tom Palmer, Rita Rasteiro.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
